system/core/tags.php

```diff
@@ -0,0 +1,192 @@
+<?php
+	// TODO:
+	// - add styling for default theme
+	// - implement error notification if new tag exists
+	// - implement tag management page: edit, translate, remove or merge tags
+	// - implement tagcontainer pagetype
+
+	function tagList(HyphaDomElement $pageListNode, $lang) {
+		/** @var HyphaDomElement $hyphaXml */
+		global $hyphaXml;
+
+		$selectedTagIds = [];
+
+		/** @var HyphaDomElement $tagEl */
+		// Build list of active tags for given page
+		foreach ($pageListNode->findXPath('tag') as $tagEl) $selectedTagIds[] = $tagEl->getAttribute('id');
+
+		$selectedTags = [];
+		$deselectedTags = [];
+
+		$lang = htmlspecialchars($lang);
+
+		// Divide list of all tags stored in hypha.xml in a selected tag list and a deselected tag list
+		foreach ($hyphaXml->findXPath('hypha/tagList/tag/language[@id=' . xpath_encode($lang) . ']') as $tagLang) {
+			/** @var HyphaDomElement $tagLang */
+			$id = $tagLang->parent()->getId();
+			$tag = ['label' => $tagLang->getAttribute('label'), 'description' => $tagLang->nodeValue];
+			if (in_array($id, $selectedTagIds)) $selectedTags[$id] = $tag;
+			else $deselectedTags[$id] = $tag;
+		}
+
+		// Construct html output
+		$html = '';
+		if (!empty($selectedTagIds) || isUser()) {
+			$html.= '<span class="prefix">' . __('tags') . ': </span>';
+			$html.= '<div class="selectedTags">';
+			foreach ($selectedTags as $id => $tag) {
+				$html.= '<div class="tagSel_'.htmlspecialchars($id).'" class="tag">';
+				$html.= '<span title="'.htmlspecialchars($tag['description']).'">'.htmlspecialchars($tag['label']).'</span>';
+				if (isUser()) $html.= ' <span class="delete-tag" title="'.__('remove-tag').'" onclick="deselectTag('.htmlspecialchars(json_encode($id)).');">⨯</span>';
+				$html.= '</div>';
+			}
+			$html.= '</div>';
+			if ((isUser() && !empty($deselectedTags)) || isAdmin()) {
+				$html.= '<select class="remainingTags" onchange="selectTag(this.value);">';
+				$html.= '<option value="_add_" selected="selected" disabled="disabled">'.__('add').'...</option>';
+				foreach ($deselectedTags as $id => $tag) {
+					$html.= '<option id="tagRem_'.htmlspecialchars($id).'" value="'.htmlspecialchars($id).'">'.htmlspecialchars($tag['label']).'</option>';
```

Comment:
What does "Rem" mean in the id here? I would think "Remove", but that's not what this select is for? Remaining then?